aptos-move/framework/aptos-stdlib/sources/math_fixed64.move

```diff
@@ -0,0 +1,144 @@
+    /// Natural log 2 in 32 bit fixed point
+    const LN2: u256 = 12786308645202655660;  // ln(2) in fixed 64 representation
+
+    /// Square root of fixed point number
+    public fun sqrt(x: FixedPoint64): FixedPoint64 {
+        let y = fixed_point64::get_raw_value(x);
+        let z = (math128::sqrt(y) << 32 as u256);
+        z = (z + ((y as u256) << 64) / z) >> 1;
+        fixed_point64::create_from_raw_value((z as u128))
+    }
+
+    /// Exponent function with a precission of 9 digits.
+    public fun exp(x: FixedPoint64): FixedPoint64 {
+        let raw_value = (fixed_point64::get_raw_value(x) as u256);
+        fixed_point64::create_from_raw_value((exp_raw(raw_value) as u128))
+    }
+
+    /// Because log2 is negative for values < 1 we instead return log2(x) + 64 which
+    /// is positive for all values of x.
+    public fun log2_plus_64(x: FixedPoint64): FixedPoint64 {
+        let raw_value = (fixed_point64::get_raw_value(x) as u128);
+        math128::log2_64(raw_value)
+    }
+
+    public fun ln_plus_32ln2(x: FixedPoint64): FixedPoint64 {
+        let raw_value = fixed_point64::get_raw_value(x);
+        let x = (fixed_point64::get_raw_value(math128::log2_64(raw_value)) as u256);
+        fixed_point64::create_from_raw_value(((x * LN2) >> 64 as u128))
+    }
+
+    /// Integer power of a fixed point number
+    public fun pow(x: FixedPoint64, n: u64): FixedPoint64 {
+        let raw_value = (fixed_point64::get_raw_value(x) as u256);
+        fixed_point64::create_from_raw_value((pow_raw(raw_value, (n as u128)) as u128))
+    }
+
+    /// Specialized function for x * y / z that omits intermediate shifting
+    public fun mul_div(x: FixedPoint64, y: FixedPoint64, z: FixedPoint64): FixedPoint64 {
+        let a = fixed_point64::get_raw_value(x);
+        let b = fixed_point64::get_raw_value(y);
+        let c = fixed_point64::get_raw_value(z);
+        fixed_point64::create_from_raw_value (math128::mul_div(a, b, c))
+    }
+
+    // Calculate e^x where x and the result are fixed point numbers
+    fun exp_raw(x: u256): u256 {
+        // exp(x / 2^64) = 2^(x / (2^64 * ln(2))) = 2^(floor(x / (2^64 * ln(2))) + frac(x / (2^64 * ln(2))))
+        let shift_long = x / LN2;
+        assert!(shift_long <= 63, std::error::invalid_state(EOVERFLOW_EXP));
+        let shift = (shift_long as u8);
+        let remainder = x % LN2;
+        // At this point we want to calculate 2^(remainder / ln2) << shift
+        // ln2 = 580 * 22045359733108027
+        let bigfactor = 22045359733108027;
+        let exponent = remainder / bigfactor;
+        let x = remainder % bigfactor;
+        // 2^(remainder / ln2) = (2^(1/580))^exponent * exp(x / 2^64)
+        let roottwo = 18468802611690918839;  // fixed point representation of 2^(1/580)
+        // 2^(1/580) = roottwo(1 - eps), so the number we seek is roottwo^exponent (1 - eps * exponent)
+        let power = pow_raw(roottwo, (exponent as u128));
+        let eps_correction = 219071715585908898;
```

Comment:
Can we add a comment about how we get the magic number?

Reply:
It's stated on line 69 no?

Reply:
I see. I was confused because l64 doesn't mention the error so I thought it's accurately 2^(1/580).

Reply:
yes integer roots of 2 are not rational (mumbles something about euclid) so we can only approximate it.